ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -180,6 +180,10 @@ class PutOperation {
   private final String loggingContext;
   private final CompressionService compressionService;
   private final ReservedMetadataIdMetrics reservedMetadataIdMetrics;
+  private PutChunk firstChunk = null;
+  private boolean firstPutChunkPrepared = false; // to track if the first chunk has been prepared for sending.
```

Comment:
I suspect that we don't need `firstPutChunkPrepared`, but only `firstPutChunkAwaitingResolution` to determine if this is a simple blob.

Reply:
I want to reduce this to only one variable/state too. I tried to think about that but wasn't able to find a reasonable solution.

Here are two problems that make 2 variables necessary in my opinion
1. The `firstChunk` is a reference to `PutChunk` object. The `PutChunk` objects are taken from the array whose size is `RouterConfig::routerMaxInMemPutChunks`. So the chunk number 5/9/13 will end up sharing the same `PutChunk` object. Thats why I think two variables are needed. `firstPutChunkAwaitingResolution` to assert that first put chunk needs to be handled and hasn't been handled yet, and `firstPutChunkPrepared ` to indicate that we don't need to worry about `firstChunk` anymore.
2. Note that `RouterConfig::routerMaxInMemPutChunks` could be just 1 or 2. In that case every chunk, or every other chunk, will basically end up being equal to `firstChunk`. So we need a way to say that `firstPutChunkNeedsResolution` (i.e, that firstChunk is read from network but we are waiting for more data to ascertain fistChunk's type) and then to say that  `firstPutChunkIsPrepared ` (i.e, first chunk is prepared, and we don't need to worry about delaying it anymore).

Reply:
I'm trying to think about a way to solve the problem with less variables and more centralized code.

For variable, 
```
we need one in PutOperation as pointer to the firstChunk. Let's call it firstChunk.
PutChunk has a variable isSimple, by default it's always false unless we change it.
```

For functions, we need modify two functions,
I guess the prepareForSending is the function we set buffer to be ready to send out to server. 
We delay the first chunk's prepareForSending until either the second chunk is ready or the chunk filling is completed.
```
private void prepareForSending {
	if (!chunkFillingCompletedSuccessfully) {
		if (chunkIndex == 0) { <--- For first chunk, just save it and return immediately.
			firstChunk = this;
			return;
		} else if (chunkIndex == 1) { <--- When the second chunk comes, complete the first chunk.
			firstChunk.prepareForSending();
			firstChunk = null;
		}
	}

	// original code as before <-- Our original code.
	...
}
```

In fillChunks, we changed two places below with <---
```
void fillChunks() {
	...
  if (chunkFillingCompletedSuccessfully) {
    PutChunk lastChunk = getBuildingChunk();
    if (lastChunk != null) {
      if (chunkCounter != 0 && lastChunk.buf.readableBytes() == 0) {
        logger.trace("{}: The last buffer(s) received from chunkFillerChannel have no data, discarding them.",
            loggingContext);
        lastChunk.releaseBlobContent();
        lastChunk.state = ChunkState.Free;
      } else {
      	lastChunk.isSimple = true; <------ Set Simple chunk to true.
        lastChunk.onFillComplete(true);
        updateChunkFillerWaitTimeMetrics();
      }
      if (lastChunk.isReady()) {
        routerCallback.onPollReady();
      }
    } else {
    	if (firstChunk != null) {
    		firstChunk.isSimple = true;
    		firstChunk.prepareForSending(); <----- Closeup the 4MB chunk.
    	}
    }
  }
}
```

Reply:
There are a few things that I want to note here
1. It's not a chunk's state that is simple or composite. It's the blob's state. For chunked uploads, each chunk is uploaded separately and gets handled like a separate put request. So if it's a chunked upload, the upload is always less than or equal to the chunk size (otherwise the upload fails). 
2. Second issue is that it's possible that `prepareForSending` was called with `chunkIndex =0`, and another chunk with `chunkIndex == 1` was created, but there was no data to fill in this chunk. In that case `prepareForSending` will not be called with `chunkIndex == 1` and `chunkFillingCompletedSuccessfully` will be set to `true`. This case needs to be handled in the `if(chunkFillingCompletedSuccessfully` section above. If we want to handle it here, then we will have to first check if `firstChunk` is still waiting (i.e., we will need at least one more variable which is `firstChunkAwaitingResolution`. 


If we want to rely on chunkCounter or chunkIndex, then I think there is another approach that can work with using `chunkCounter` instead of `chunkIndex` and checking the size of the buffer if `chunkCounter == 1`. Let me try to update the PR with that.

Reply:
1. Understood. That's why I  set isSimple to true only if it's not chunked Upload as below.
  ```
  lastChunk.isSimple = true if it's not chunkUpload; 
  ```
2. In the above code, I added the condition under chunkFillingCompletedSuccessfully to handle that case.

Reply:
But `isSimple` is not a property of a chunk. Every is chunk is always "simple" because its always <=4Mib. `isSimple` is the property of a user blob.

Reply:
Also there is one more issue why I don't prefer doing `firstChunk.onFillComplete`  inside the `onFillCompelete` method. The reason is that, for all large blobs (larger than 8MiB), we will need to wait for reading the entire 4MiB of the second chunk, before we can send the first chunk to the server. This shouldn't increase the upload time overall, but it's more efficient to send the first chunk to the server, as soon as we have at least one more chunk on the network.

Reply:
There are two more things to note in the above code snippet:
1. When we set the `firstChunk` variable in the `prepareForSending` method, then the `firstChunk` is still in `Building` state. When we need a new chunk for more data coming from the network, we will call `getBuildingChunk`, which will be the first chunk. So the above logic also needs change in `getBuildingChunk` to make it work.

2. Even if we make the change in `getBuildingChunk` as mentioned above, note that it's possible that `lastChunk != null`, and firstChunk is still waiting. Because it can happen that `firstChunk` was filled, and `prepareForSending` was called. But there we didn't really send it because we want to wait for more data. After some time we got an empty buffer on the network. And later `chunkFillingCompletedSuccessfully` was marked as `true`, without reading any more data. We will need to find a way to handle this case as well in the above logic.

Reply:
> But `isSimple` is not a property of a chunk. Every is chunk is always "simple" because its always <=4Mib. `isSimple` is the property of a user blob.

Although isSimple is blob concept, we represent it on the CHUNK level in the prepareForSending, right?
I agree we can move the variable to the PutOperation instead inside the PutChunk.
```
        BlobDataType blobDataType =
            isMetadataChunk() ? BlobDataType.METADATA : isSimpleBlob ? BlobDataType.SIMPLE : 
```

Reply:
> Also there is one more issue why I don't prefer doing `firstChunk.onFillComplete` inside the `onFillCompelete` method. The reason is that, for all large blobs (larger than 8MiB), we will need to wait for reading the entire 4MiB of the second chunk, before we can send the first chunk to the server. This shouldn't increase the upload time overall, but it's more efficient to send the first chunk to the server, as soon as we have at least one more chunk on the network.

I don't have much concern about it. For memory usage, we will use at least two chunks for large blob anyway. And the wait only happens on the second chunk, not the following chunks. For large blob, there wont be no difference.
For 8MB chunk, maybe there is a little bit difference but can be ignored.

Reply:
To be precise, what we say in the `blobDataType` is that whether this chunk belongs to a larger blob, or is this a blob in itself. In order to make that determination, we need to know the state of the blob.
Consider this in another way. If a blob has multiple data chunks, all the chunks will need to be `chunk.isSimple = false`. If the blob has only one chunk, then all the chunks (in this case only one chunk) will need to be `chunk.isSimple = true`. So having it at a chunk level doesn't help.

Reply:
> I don't have much concern about it
My point is, if there is a way to keep it performant, then why not. Also, that's not my main concern. The change in itself will need more changes anyways.

Let's discuss these changes offline, and I will update the comments with the discussion summary.

Reply:
> 

1. Good point for the concern of the getBuildingChunk. We can solve it by adding another state before READY. Let's say RESOLVING. 
2. Solve the problem 2. 

Modified my proposal. Just to share my thought. 
You mentioned you have another version. Happy to review that version. 

Variables:
```
PutOperation: Add two variables
   boolean isSimpleBlob = !options.isChunkUpload();
   boolean blobTypeIsResolved = false;
Chunk State: Add one state
   Add one chunk state: RESOLVING.
```

Code change to two functions:
```
private void prepareForSending {
        if (chunkIndex == 1) { <--- When the second chunk comes, set all the flags.
                isSimpleBlob = false;
                blobTypeIsResolved = true;
                resolveWaitingChunk();
	}
	if (!blobTypeIsResolved) {
             state = ChunkState.RESOLVING;
             return;
	}

	// original code as before
	...
}
```

```
void fillChunks() {
	...
      if (chunkFillingCompletedSuccessfully) {
        PutChunk lastChunk = getBuildingChunk();
        if (lastChunk != null) {
           ....
        }
        blobTypeIsResolved = true; <----- 
         resolveWaitingChunk(); <----- Check any chunk which is waiting for resolving.
  }
}
```

```
void resolveWaitingChunk(){
         for (PutChunk chunk : putChunks) { <--- Check if we have a first chunk haven't been resolved.
             if (chunk.state == RESOLVING()) {
                prepareForSending();
                break;
             }
          }
}
```